Compile failed with aarch64-linux-android

when I compile wasmtime for android using: ```cargo build --target  aarch64-linux-android --release```

The output is: 
```
error[E0432]: unresolved import `crate::dir::SeekLoc`
 

--> crates/wasi-common/yanix/src/sys/mod.rs:1:5
  |
1 | use crate::dir::SeekLoc;
  |     ^^^^^^^^^^^^^^^^^^^ no `SeekLoc` in `dir`

error[E0432]: unresolved import `crate::dir::SeekLoc`
 --> crates/wasi-common/yanix/src/sys/linux/dir.rs:1:40
  |
1 | use crate::dir::{Dir, Entry, EntryExt, SeekLoc};
  |                                        ^^^^^^^ no `SeekLoc` in `dir`

error[E0425]: cannot find function `futimes` in crate `libc`
    --> crates/wasi-common/yanix/src/sys/linux/utimesat.rs:30:38
     |
30   |     from_success_code(unsafe { libc::futimes(f.as_raw_fd(), times.as_ptr()) })
     |                                      ^^^^^^^ help: a function with a similar name exists: `futimens`
     |
    ::: /Users/niexinxin/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.72/src/unix/linux_like/mod.rs:1319:5
     |
1319 |     pub fn futimens(fd: ::c_int, times: *const ::timespec) -> ::c_int;
     |     ------------------------------------------------------------------ similarly named function `futimens` defined here

```

--- a/crates/wasi-common/src/sys/unix/mod.rs
+++ b/crates/wasi-common/src/sys/unix/mod.rs
@@ -9,7 +9,8 @@ pub(crate) mod poll;
 pub(crate) mod stdio;
 
 cfg_if::cfg_if! {
-    if #[cfg(target_os = "linux")] {
+    if #[cfg(any(target_os = "linux",
+                 target_os = "android"))] {
         mod linux;
         use linux as sys_impl;
     } else if #[cfg(target_os = "emscripten")] {

--- a/crates/wasi-common/yanix/src/dir.rs
+++ b/crates/wasi-common/yanix/src/dir.rs
@@ -36,9 +36,15 @@ impl Dir {
     }
 
     /// Set the position of the directory stream, see `seekdir(3)`.
-    #[cfg(not(target_os = "android"))]
     pub fn seek(&mut self, loc: SeekLoc) {
-        unsafe { libc::seekdir(self.0.as_ptr(), loc.0) }
+        // https://github.com/rust-lang/libc/pull/1996
+        #[cfg(not(target_os = "android"))]
+        use libc::seekdir;
+        #[cfg(target_os = "android")]
+        extern "C" {
+            fn seekdir(dirp: *mut libc::DIR, loc: libc::c_long);
+        }
+        unsafe { seekdir(self.0.as_ptr(), loc.0) }
     }
 
     /// Reset directory stream, see `rewinddir(3)`.
@@ -50,10 +56,16 @@ impl Dir {
     ///
     /// If this location is given to `Dir::seek`, the entries up to the previously returned
     /// will be omitted and the iteration will start from the currently pending directory entry.
-    #[cfg(not(target_os = "android"))]
     #[allow(dead_code)]
     pub fn tell(&self) -> SeekLoc {
-        let loc = unsafe { libc::telldir(self.0.as_ptr()) };
+        #[cfg(not(target_os = "android"))]
+        use libc::telldir;
+        #[cfg(target_os = "android")]
+        extern "C" {
+            fn telldir(dirp: *mut libc::DIR) -> libc::c_long;
+        }
+        // https://github.com/rust-lang/libc/pull/1996
+        let loc = unsafe { telldir(self.0.as_ptr()) };
         SeekLoc(loc)
     }
 
@@ -93,11 +105,9 @@ impl Entry {
     }
 }
 
-#[cfg(not(target_os = "android"))]
 #[derive(Clone, Copy, Debug)]
 pub struct SeekLoc(pub(crate) libc::c_long);
 
-#[cfg(not(target_os = "android"))]
 impl SeekLoc {
     pub fn to_raw(&self) -> i64 {
         self.0.into()

--- a/crates/wasi-common/yanix/src/file.rs
+++ b/crates/wasi-common/yanix/src/file.rs
@@ -90,11 +90,26 @@ bitflags! {
         const WRONLY = libc::O_WRONLY;
         const RDWR = libc::O_RDWR;
         #[cfg(any(target_os = "linux",
+                  target_os = "android",
                   target_os = "netbsd",
                   target_os = "openbsd",
                   target_os = "wasi",
                   target_os = "emscripten"))]
-        const RSYNC = libc::O_RSYNC;
+        const RSYNC = {
+            // Have to use cfg_if: https://github.com/bitflags/bitflags/issues/137
+            cfg_if! {
+                if #[cfg(any(target_os = "linux",
+                             target_os = "netbsd",
+                             target_os = "openbsd",
+                             target_os = "wasi",
+                             target_os = "emscripten"))] {
+                    libc::O_RSYNC
+                } else if #[cfg(target_os = "android")] {
+                    // Android defines O_RSYNC as O_SYNC
+                    libc::O_SYNC
+                }
+            }
+        };
         const SYNC = libc::O_SYNC;
         const TRUNC = libc::O_TRUNC;
         #[cfg(any(target_os = "linux",

--- a/crates/wasi-common/yanix/src/sys/linux/filetime.rs
+++ b/crates/wasi-common/yanix/src/sys/linux/filetime.rs
@@ -53,5 +53,8 @@ pub fn utimensat(
         return Err(err);
     }
 
-    super::utimesat::utimesat(dirfd, path, atime, mtime, symlink_nofollow)
+    #[cfg(not(target_os = "android"))]
+    return super::utimesat::utimesat(dirfd, path, atime, mtime, symlink_nofollow);
+    #[cfg(target_os = "android")]
+    unreachable!();
 }

--- a/crates/wasi-common/yanix/src/sys/linux/mod.rs
+++ b/crates/wasi-common/yanix/src/sys/linux/mod.rs
@@ -2,4 +2,5 @@ pub(crate) mod dir;
 pub(crate) mod fadvise;
 pub(crate) mod file;
 pub(crate) mod filetime;
+#[cfg(not(target_os = "android"))]
 pub(crate) mod utimesat;

